ttl: ensure each DistSQL processor only deletes its own data

ecwall opened the pull request:
refs https://github.com/cockroachdb/cockroach/issues/85800
    
Fix span/range calculation to filter out rows owned by other nodes.

Rows are grouped by node via DistSQLPlanner.PartitionSpans instead
of RangeIterator. Therefore, rows in the same range but a different table
or index no longer need to be filtered out.

Release note: None

Release justification: TTL performance bug fix.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86057)
<!-- Reviewable:end -->

ecwall commented on pkg/sql/ttl/ttljob/ttljob_processor.go:
```diff
@@ -131,27 +126,16 @@ func (ttl *ttlProcessor) work(ctx context.Context) error {
 				for rangeToProcess := range rangeChan {
 					start := timeutil.Now()
-					rangeRowCount, err := runTTLOnRange(
+					rangeRowCount, err := ttl.runTTLOnRange(
```
Changed this to have a `ttlProcessor` receiver to reduce the number of params.

ecwall commented on pkg/sql/ttl/ttljob/ttljob_processor.go:
```diff
@@ -224,57 +208,63 @@ func (ttl *ttlProcessor) work(ctx context.Context) error {
+	return serverCfg.JobRegistry.UpdateJobWithTxn(
+		ctx,
+		jobID,
+		nil,  /* txn */
```
It creates a transaction internally if nil.

ecwall commented on pkg/sql/ttl/ttljob/ttljob_processor.go:
```diff
@@ -224,57 +208,63 @@ func (ttl *ttlProcessor) work(ctx context.Context) error {
+		ctx,
+		jobID,
+		nil,  /* txn */
+		true, /* useReadLock */
```
This prevents the race condition causing the test flakiness.

ecwall commented on pkg/sql/ttl/ttljob/ttljob_test.go:
```diff
@@ -127,6 +129,9 @@ func newRowLevelTTLTestJobTestHelper(
 func (h *rowLevelTTLTestJobTestHelper) waitForScheduledJob(
 	t *testing.T, expectedStatus jobs.Status, expectedErrorRe string,
 ) {
+	h.env.SetTime(timeutil.Now().Add(time.Hour * 24))
```
These 2 lines were always called before `waitForScheduledJob` so I'm inlining them here to reduce duplication.

ecwall commented on pkg/sql/ttl/ttljob/ttljob_test.go:
```diff
@@ -299,17 +337,69 @@ INSERT INTO t (id, crdb_internal_expiration) VALUES (1, now() - '1 month'), (2,
 	}
 }
 
+func TestRowLevelTTLJobMultipleNodes(t *testing.T) {
```
I changed this to its own test because the impl is very different from the other subtests.

ecwall commented on pkg/sql/ttl/ttljob/ttljob_test.go:
```diff
@@ -299,17 +337,69 @@ INSERT INTO t (id, crdb_internal_expiration) VALUES (1, now() - '1 month'), (2,
+		sqlDB.Exec(t, `INSERT INTO tbl VALUES ($1, $2)`, splitAt+i+1, expiredTs)
+		expectedNumExpiredRows += 2
+		ranges := sqlDB.QueryStr(t, `SHOW RANGES FROM TABLE tbl`)
+		if ranges[0][4] != ranges[1][4] {
```
This seems kind of hacky.  Is there a better way to ensure multiple leaseholders?

ecwall commented on pkg/sql/ttl/ttljob/ttljob_processor.go:
It looks like the nodes can select outside of the range they are leaseholders for so I need to fix that, but I'll do it separately.

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86057)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XUtff8ITgEvXFia4m:-N9XUtff8ITgEvXFia4n:bae9jrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L220)):*
> ```Go
> 			progress := md.Progress
> 			existingRowCount := progress.Details.(*jobspb.Progress_RowLevelTTL).RowLevelTTL.RowCount
> 			progress.Details.(*jobspb.Progress_RowLevelTTL).RowLevelTTL.RowCount += processorRowCount
> ```

should this use `atomic.LoadInt64` first to access the processorRowCount?
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XV16e6wdnda4SrFup:-N9XV16e6wdnda4SrFuq:b-skfz7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L222)):*
> ```Go
> 			progress.Details.(*jobspb.Progress_RowLevelTTL).RowLevelTTL.RowCount += processorRowCount
> 			ju.UpdateProgress(progress)
> 			log.Infof(
> ```

how frequently will this log? it feels like it could be something that should only be shown with `log.Vinfof(ctx, 2, ...)`
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XVmFz-zHL9UlebnvB:-N9XVmFz-zHL9UlebnvC:btyo2h3) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L233)):*
> ```Go
> 
> // rangeRowCount should be checked even if the function returns an error because it may have partially succeeded
> func (ttl *ttlProcessor) runTTLOnRange(
> ```

nit: function receivers are usually one character by convention. otherwise this is easily confused with the `ttl` package. (and use the same for all the functions that have this receiver.) i'd say use either `t` or `p`.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9Ik4tKVWdaZLWGeVbT-r1-302:-N9XWSCB4t0vaecJJHJq:b-s1jlco) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L302)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

It looks like the nodes can select outside of the range they are leaseholders for so I need to fix that, but I'll do it separately.
</blockquote></details>

this should be a TODO in the code, not just a comment on the PR
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XWA9PBRpeHLP4XBCI:-N9XWA9PBRpeHLP4XBCJ:b-9swk1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L303)):*
> ```Go
> 		expiredRowsPKs, err := selectBuilder.run(ctx, ie)
> 		log.Dev.Infof(
> 			ctx,
> ```

nit: use `log.Vinfof(ctx, 2, ...)`
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XWCPB-WsTAxRFzhaU:-N9XWCPB-WsTAxRFzhaV:b-cjx5a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L305)):*
> ```Go
> 			ctx,
> 			"TTL selected rows jobID=%d processorID=%d tableID=%d expiredRowsPKs=%s",
> 			ttlSpec.JobID, ttl.ProcessorID, details.TableID, expiredRowsPKs,
> ```

how big is `expiredRowsPKs`? the log won't be useful if this prints out hundreds/thousands of entries.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86057)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XUtff8ITgEvXFia4m:-N9Xa9kjB7ZNSzy-aqAb:bbkr6mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L220)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should this use `atomic.LoadInt64` first to access the processorRowCount?
</blockquote></details>

`processorRowCount` isn't being modified anymore by this point so there isn't a race between updates and reads.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XV16e6wdnda4SrFup:-N9XaQdm2zZ-8qaPUIjQ:bjxwtpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L222)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

how frequently will this log? it feels like it could be something that should only be shown with `log.Vinfof(ctx, 2, ...)`
</blockquote></details>

It will log once per leaseholder per TTL run. I can change it to that though.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XVmFz-zHL9UlebnvB:-NA5_X-wDtf1-JCLZyj1:bgebuba) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L233)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: function receivers are usually one character by convention. otherwise this is easily confused with the `ttl` package. (and use the same for all the functions that have this receiver.) i'd say use either `t` or `p`.
</blockquote></details>

Changed to `t`.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9Ik4tKVWdaZLWGeVbT-r1-302:-NA5_ZyR-hbSiYQT_WRY:beyqzz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L302)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this should be a TODO in the code, not just a comment on the PR
</blockquote></details>

Addressed in prior PR
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XWA9PBRpeHLP4XBCI:-NA5_chqCqvwulijoInE:b-xajej7) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L303)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: use `log.Vinfof(ctx, 2, ...)`
</blockquote></details>

Addressed in prior PR.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-N9XWCPB-WsTAxRFzhaU:-NA5_e0Q2U-NR2D3SMrh:bo8ydvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/153d6e192fe793bb00d56ac41c731421ca5d6d63/pkg/sql/ttl/ttljob/ttljob_processor.go#L305)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

how big is `expiredRowsPKs`? the log won't be useful if this prints out hundreds/thousands of entries.
</blockquote></details>

Removed log.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86057)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 157 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-NA5_wOP7ozNaZV2_owj:-NA5_wOP7ozNaZV2_owk:b2puazh) ([raw file](https://github.com/cockroachdb/cockroach/blob/481b537685e5091941fcbcf8757059238b27d7ab/pkg/sql/ttl/ttljob/ttljob_processor.go#L157)):*
> ```Go
> 		ri := kvcoord.MakeRangeIterator(serverCfg.DistSender)
> 		for _, span := range ttlSpec.Spans {
> 			spanStartKey := span.Key
> ```

This logic works correctly in the multi-node test, but the single node tests are failing.  I don't quite understand the relationship between a span and a range and need help with this logic. 


<!-- Sent from Reviewable.io -->

ecwall committed:
ttl: ensure each DistSQL processor only deletes its own data

refs https://github.com/cockroachdb/cockroach/issues/85800

Fix span/range calculation to filter out rows owned by other nodes.

Rows are grouped by node via DistSQLPlanner.PartitionSpans instead
of RangeIterator. Therefore, rows in the same range but a different table
or index no longer need to be filtered out.

Release note: None

Release justification: TTL performance bug fix.

yuzefovich commented:
I think we need to backport fixes to the key decoding to 22.1 if TTL feature is enabled for tenants in serverless.

rafiss reviewed (changes requested):
nice, i just had questions since i don't fully understand the fix

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-:-NAQKma752aujZ-y5j8M:boz0v9y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 153 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-NAQK0BX1-sZDmRwXlSo:-NAQK0BX1-sZDmRwXlSp:b-fe9f56) ([raw file](https://github.com/cockroachdb/cockroach/blob/3348aaa043fdf277576cdc0f1ed8e6c7aee5bbf1/pkg/sql/ttl/ttljob/ttljob_processor.go#L153)):*
> ```Go
> 		// Iterate over every range to feed work for the goroutine processors.
> 		var alloc tree.DatumAlloc
> 		for _, span := range ttlSpec.Spans {
> ```

interesting fix -- why don't we need the rangeIterator any more? could you explain in the commit message?
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 374 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-NAQKE_c9oiT6T8D4UjO:-NAQKE_c9oiT6T8D4UjP:bptfsf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0ae4288ae19059f5622b0679e3622e94f3c503c/pkg/sql/ttl/ttljob/ttljob_processor.go#L374)):*
> ```Go
> 	rKey := key.AsRawKey()
> 
> 	// If any of these errors, that means we reached an "empty" key, which
> ```

i didn't follow why these two checks should be removed. could you explain in the commit message?
___
*[`pkg/sql/ttl/ttljob/ttljob_test.go` line 198 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-NAQK_lHALQnB42PZFIn:-NAQK_lHALQnB42PZFIo:b-ardaae) ([raw file](https://github.com/cockroachdb/cockroach/blob/3348aaa043fdf277576cdc0f1ed8e6c7aee5bbf1/pkg/sql/ttl/ttljob/ttljob_test.go#L198)):*
> ```Go
> }
> 
> // todo(ewall): migrate usages to verifyExpiredRows and switch SPLIT AT usage to SplitTable
> ```

i thought we already did the SplitTable change?


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86057)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 153 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-NAQK0BX1-sZDmRwXlSo:-NAQQaLq3wYv96pOmpzA:byvqlcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3348aaa043fdf277576cdc0f1ed8e6c7aee5bbf1/pkg/sql/ttl/ttljob/ttljob_processor.go#L153)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

interesting fix -- why don't we need the rangeIterator any more? could you explain in the commit message?
</blockquote></details>

I think the range iterator was to group the batches by range to make them more efficient, but `ttlSpec.Spans` (sourced from `DistSQLPlanner.PartitionSpans`) already only has the spans relevant to this node.

I think in the future we can remove the parallel `Span` processing and increase the batch size to reduce the number of config params, but I need to test this in the workload.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 374 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-NAQKE_c9oiT6T8D4UjO:-NAQRnYY5bw5dNPh5U5Z:b-o2iqis) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0ae4288ae19059f5622b0679e3622e94f3c503c/pkg/sql/ttl/ttljob/ttljob_processor.go#L374)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i didn't follow why these two checks should be removed. could you explain in the commit message?
</blockquote></details>

I don't think these branches were being hit anymore. @otan said tests were previously failing without them, but now the tests pass without them: https://cockroachlabs.slack.com/archives/C01RX2G8LT1/p1661281166206439

I'll update the message.
___
*[`pkg/sql/ttl/ttljob/ttljob_test.go` line 198 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-NAQK_lHALQnB42PZFIn:-NAQS71c7WOxrnoqgfWw:bjscj9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3348aaa043fdf277576cdc0f1ed8e6c7aee5bbf1/pkg/sql/ttl/ttljob/ttljob_test.go#L198)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i thought we already did the SplitTable change?
</blockquote></details>

I did for the multi-node tests, but other tests were failing because of some Voter error so I need to look into it more.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice, thanks for the explanations

> Reviewed 3 of 4 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86057#-:-NAehiFN7sYkmpPp1jOM:b9tvjxd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)



<!-- Sent from Reviewable.io -->

ecwall commented:
bors r=rafiss
